jbi/queue.py

```diff
@@ -0,0 +1,315 @@
+    error: Optional[PythonException] = None
+
+    @property
+    def identifier(self):
+        return f"{self.timestamp.isoformat()}-{self.payload.bug.id}-{self.payload.event.action}-{"error" if self.error else "postponed"}"
+
+
+@lru_cache(maxsize=1)
+def get_dl_queue():
+    settings = get_settings()
+    return DeadLetterQueue(settings.dl_queue_dsn)
+
+
+class QueueBackend(ABC):
+    """An interface for dead letter queues."""
+
+    @abstractmethod
+    def ping(self):
+        """Report if the queue backend is available and ready to be written to"""
+        pass
+
+    @abstractmethod
+    async def clear(self):
+        """Remove all bugs and their items from the queue"""
+        pass
+
+    @abstractmethod
+    async def put(self, item: QueueItem):
+        """Insert item into queued items for a bug, maintaining sorted order by
+        payload event time ascending
+        """
+        pass
+
+    @abstractmethod
+    async def remove(self, bug_id: int, identifier: str):
+        """Remove an item from the target bug's queue. If the item is the last
+        one for the bug, remove the bug from the queue entirely.
+        """
+        pass
+
+    @abstractmethod
+    async def get(self, bug_id: int) -> list[QueueItem]:
+        """Retrieve all of the queue items for a specific bug, sorted in
+        ascending order by the timestamp of the payload event.
+        """
+        pass
+
+    @abstractmethod
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        """Retrive all items in the queue, grouped by bug
+
+        Returns:
+            dict[int, list[QueueItem]]: Returns a dict of
+            {bug_id: list of events}. Each list of events sorted in ascending
+            order by the timestamp of the payload event.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def size(self) -> int:
+        pass
+
+
+class MemoryBackend(QueueBackend):
+    def __init__(self):
+        self.existing: dict[int, list[QueueItem]] = defaultdict(list)
+
+    def ping(self):
+        return True
+
+    async def clear(self):
+        logger.debug("Clearing queue")
+        self.existing.clear()
+
+    async def put(self, item: QueueItem):
+        bisect.insort_right(
+            self.existing[item.payload.bug.id],
+            item,
+            key=lambda i: i.payload.event.time or 0,
+        )
+
+    async def get(self, bug_id: int) -> list[QueueItem]:
+        # though we're using a defaultdict, check if the bug is in the queue
+        # this way so that we don't create a key for a bug with no items
+        if bug_id not in self.existing:
+            return []
+        items = self.existing[bug_id]
+        if not items:
+            logger.warn("No items for bug %s, but present in queue", bug_id)
+        return items
+
+    async def get_all(self) -> dict[int, list[QueueItem]]:
+        empty_entries = [not items for items in self.existing.values]
+        if empty_entries:
```

Comment:
You could use `if empty_entries := [...]`: 😊

Reply:
[9bd9c96](https://github.com/mozilla/jira-bugzilla-integration/pull/918/commits/9bd9c96561d8d3149310037c84a26227e5c80044)